From d3fb08ae299e5354d910af4e5ef51bc7a8138df9 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Fri, 3 Nov 2023 13:25:11 +0530
Subject: [PATCH] wifi: cfg80211: fix off channel operation allowed check for
 MLO

Currently, in off channel operation allowed check, if any link in the
interface operates on a RADAR channel, the check returns false under
the assumption that RADAR channel can not be left for off channel
scanning. But this assumption is true only for non-MLO case.

This is because, with MLO, there could be multiple links associated
with the interface and scan could happen in other underlying HW. Hence,
it can do a off channel scan.

Add fix to check if there is a requested channel on the HW which already
operate on the RADAR channel, and if so then return false for the off
channel operation check or else true.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 net/wireless/nl80211.c | 26 ++++++++++++++++++++------
 1 file changed, 20 insertions(+), 6 deletions(-)

diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index e720b2dd30f6..0b3b8d9e5ded 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -9605,11 +9605,19 @@ static bool cfg80211_off_channel_oper_allowed(struct wireless_dev *wdev,
 {
 	unsigned int link_id;
 	bool all_ok = true;
+	int req_hw_idx;
 
 	ASSERT_WDEV_LOCK(wdev);
 
 	if (!cfg80211_beaconing_iface_active(wdev))
-		return true;
+		return all_ok;
+
+	if (WARN_ON(!chan))
+		return !all_ok;
+
+	req_hw_idx = cfg80211_get_hw_idx_by_chan(wdev->wiphy, chan);
+	if (WARN_ON(req_hw_idx == -1))
+		return !all_ok;
 
 	/*
 	 * FIXME: check if we have a free HW resource/link for chan
@@ -9621,19 +9629,25 @@ static bool cfg80211_off_channel_oper_allowed(struct wireless_dev *wdev,
 	/* we cannot leave radar channels */
 	for_each_valid_link(wdev, link_id) {
 		struct cfg80211_chan_def *chandef;
+		int link_hw_idx;
 
 		chandef = wdev_chandef(wdev, link_id);
 		if (!chandef || !chandef->chan)
 			continue;
 
+		if (!(chandef->chan->flags & IEEE80211_CHAN_RADAR))
+			continue;
+
 		/*
-		 * FIXME: don't require all_ok, but rather check only the
-		 *	  correct HW resource/link onto which 'chan' falls,
-		 *	  as only that link leaves the channel for doing
-		 *	  the off-channel operation.
+		 * okay so channel is radar channel. If the HW resource/link onto
+		 * which 'chan' falls is the same HW resource/link onto which this
+		 * channel falls then it will go off-channel
 		 */
+		link_hw_idx = cfg80211_get_hw_idx_by_chan(wdev->wiphy, chandef->chan);
+		if (WARN_ON(link_hw_idx == -1))
+			all_ok = false;
 
-		if (chandef->chan->flags & IEEE80211_CHAN_RADAR)
+		if (link_hw_idx == req_hw_idx)
 			all_ok = false;
 	}
 
-- 
2.17.1

